lib/linked_in/assets.rb

```diff
@@ -0,0 +1,130 @@
+require 'mime/types'
+require 'open-uri'
+
+module LinkedIn
+  DEFAULT_TIMEOUT_SECONDS = 300
+  POLL_SLEEP_SECONDS = 5
+
+  CONTENT_TYPE = "application/json"
+  UPLOAD_MECHANISM = "com.linkedin.digitalmedia.uploading.MediaUploadHttpRequest"
+
+  # Assets APIs
+  #
+  # @see https://docs.microsoft.com/en-us/linkedin/marketing/integrations/community-management/shares/vector-asset-api
+  #
+  # [(contribute here)](https://github.com/mdesjardins/linkedin-v2)
+  class Assets < APIResource
+    class UploadFailed < StandardError; end
+    class UploadTimeout < UploadFailed; end
+    class UploadIncomplete < UploadFailed; end
+    class UploadClientError < UploadFailed; end
+
+    # Upload images and videos to LinkedIn from a supplied URL.
+    #
+    # @see https://docs.microsoft.com/en-us/linkedin/marketing/integrations/community-management/shares/vector-asset-api#upload-the-image
+    #
+    # @options options [String] :owner, the owner entity id.
+    # @options options [String] :source_url, the URL to the content to be uploaded.
+    # @options options [Numeric] :timeout, optional timeout value in seconds, defaults to 300.
+    # @return [String], the asset entity
+    #
+    def upload(options = {})
+      asset_entity, upload_url = register_upload(options)
+
+      source_url = options.delete(:source_url)
+      timeout = options.delete(:timeout) || DEFAULT_TIMEOUT_SECONDS
+
+      media_upload_endpoint = upload_url
+
+      media = open(source_url, 'rb')
```

Comment:
This has been a thing on the media endpoint since forever but this probably isn't safe if users are ever allowed to pass a source url to this. Pretty sure they could pass a local path and read anything on the filesystem. I don't think it's worth fixing here right now but something to keep in mind when we're using it.